packages/theme-language-server-common/src/documents/DocumentManager.ts

```diff
@@ -93,15 +118,39 @@ export class DocumentManager {
    *
    * Files that are loaded from the AbstractFileSystem will have a version of `undefined`.
    */
-  public async preload(rootUri: URI) {
-    if (!this.fs) throw new Error('Cannot call preload without a FileSystem');
-    const missingFiles = await recursiveReadDirectory(
-      this.fs,
-      rootUri,
-      ([uri]) => /.(liquid|json)$/.test(uri) && !this.sourceCodes.has(uri),
-    );
-    await Promise.all(
-      missingFiles.map(async (file) => this.set(file, await this.fs!.readFile(file), undefined)),
-    );
-  }
+  public preload = memoize(
+    async (rootUri: UriString) => {
+      if (!this.fs) throw new Error('Cannot call preload without a FileSystem');
+      const { fs, connection, clientCapabilities } = this;
+      const progress = Progress.create(connection, clientCapabilities, `preload#${rootUri}`);
+
+      progress.start('Initializing Liquid LSP');
+
+      const missingFiles = await recursiveReadDirectory(
```

Comment:
Maybe an opportunity to rename this variable that has confused me a couple of times:

```suggestion
      const filesToLoad = await recursiveReadDirectory(
```